torch/csrc/jit/codegen/cuda/root_domain_map.h

```diff
@@ -0,0 +1,342 @@
+  //! \param id_b An IterDomain in td_b
+  //! \returns Boolean representing if they are mapped
+  bool canMap(
+      const TensorDomain* td_a,
+      const IterDomain* id_a,
+      const TensorDomain* td_b,
+      const IterDomain* id_b) const;
+
+  //! Make a TensorDomain an alias of another TensorDomain
+  //!
+  //! This is for the computeAt transformation, where TensorViews are
+  //! updated with new TensorDomains. Since they keep using the same
+  //! root doamins, the root mapping remains valid but needs to
+  //! reflect the use of new TensorDomains as aliases of the existing
+  //! ones.
+  //!
+  //! \param td An existing TensorDomain
+  //! \param td_alias An alias of td
+  void setAlias(const TensorDomain* td, const TensorDomain* td_alias);
+
+ private:
+  //! Returns if key_a and key(td_b, id_b) are mapped to eachother (equivalent),
+  //! or are the same key.
+  //!
+  //! \param key_a A DomainKey
+  //! \param td_b Another TensorDomain
+  //! \param id_b An IterDomain in td_b
+  //! \returns Boolean representing if they are mapped
+  bool canMap(
+      const DomainKey& key_a,
+      const TensorDomain* td_b,
+      const IterDomain* id_b) const;
+
+  //! Returns if key_a and key_b are mapped to eachother (equivalent), or are
+  //! the same key.
+  bool canMap(const DomainKey& key_a, const DomainKey& key_b) const;
+
+  //! Returns the set of (non-broadcast) DomainKeys that id in td is
+  //! broadcasted to. Can result in more than one "concrete" DomainKey.
+  std::vector<DomainKey> getConcretizedKeys(
+      const TensorDomain* td,
+      const IterDomain* id) const;
+
+  //! Returns the set of (non-broadcast) iter domains that id in td is
+  //! broadcasted to. Can result in more than one "concrete" iter domain.
+  std::unordered_set<const IterDomain*>& getConcretizedDomains(
+      const TensorDomain* td,
+      const IterDomain* id);
+
+  //! Return a map between root IterDomains of a producer-consumer
+  //! pair.
+  //!
+  //! \param producer A producer TensorDomain
+  //! \param consumer A consumer TensorDomain
+  //! \param root_dims_to_map Maps only from IterDomains in this set
+  //! \param producer_to_consumer Maps from producer to consumer if true
+  std::unordered_map<IterDomain*, IterDomain*> map(
+      const TensorDomain* producer,
+      const TensorDomain* consumer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map,
+      bool producer_to_consumer) const override;
+
+ private:
+  //! Disjoint set of all mapped <TD, ID> keys to determine axes equivalency
+  DisjointSet<DomainKey, DomainKeyHash> eq_set_;
+
+  //! All IterDomains in the mapping that are a broadcast ID
+  DomainKeyMap<std::unordered_set<const IterDomain*>> bcast_map_;
+
+  //! Broadcast iter domain that does not match dimensions in its produer,
+  //! meaning it is a brand new domain in its TensorDomain.
+  DomainKeySet new_broadcast_domains_;
+
+  friend std::string toString(const ComputeAtRootDomainMap&);
```

Comment:
nit: please move the friend declaration to the top of the struct/class definition

Reply:
Thanks. Done.